cpp/include/cuml/experimental/fil/detail/gpu_introspection.hpp

```diff
@@ -0,0 +1,167 @@
+ * limitations under the License.
+ */
+#pragma once
+#include <cuda_runtime_api.h>
+#include <cuml/experimental/fil/detail/index_type.hpp>
+#include <cuml/experimental/fil/detail/raft_proto/cuda_check.hpp>
+#include <cuml/experimental/fil/detail/raft_proto/device_id.hpp>
+#include <cuml/experimental/fil/detail/raft_proto/device_type.hpp>
+#include <vector>
+
+namespace ML {
+namespace experimental {
+namespace fil {
+namespace detail {
+
+inline auto get_max_shared_mem_per_block(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
+  auto static cache = std::vector<int>{};
+  if (cache.size() == 0) {
+    auto device_count = int{};
+    raft_proto::cuda_check(cudaGetDeviceCount(&device_count));
+    cache.resize(device_count);
+    for (auto dev=0; dev < device_count; ++dev) {
+      raft_proto::cuda_check(
+        cudaDeviceGetAttribute(
+          &(cache[dev]),
+          cudaDevAttrMaxSharedMemoryPerBlockOptin,
+          dev
+        )
+      );
+    }
+  }
+  return index_type(cache.at(device_id.value()));
+}
+
+inline auto get_sm_count(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
+  auto static cache = std::vector<int>{};
+  if (cache.size() == 0) {
+    auto device_count = int{};
+    raft_proto::cuda_check(cudaGetDeviceCount(&device_count));
+    cache.resize(device_count);
+    for (auto dev=0; dev < device_count; ++dev) {
+      raft_proto::cuda_check(
+        cudaDeviceGetAttribute(
+          &(cache[dev]),
+          cudaDevAttrMultiProcessorCount,
+          dev
+        )
+      );
+    }
+  }
+  return index_type(cache.at(device_id.value()));
+}
+
+inline auto get_max_threads_per_block(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
+  auto result = int{};
+  raft_proto::cuda_check(
+    cudaDeviceGetAttribute(
+      &result,
+      cudaDevAttrMaxThreadsPerBlock,
+      device_id.value()
+    )
+  );
+  return index_type(result);
+}
+
+inline auto get_max_threads_per_sm(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
+  auto result = int{};
+  raft_proto::cuda_check(
+    cudaDeviceGetAttribute(
+      &result,
+      cudaDevAttrMaxThreadsPerMultiProcessor,
+      device_id.value()
+    )
+  );
+  return index_type(result);
+}
+
+inline auto get_max_shared_mem_per_sm(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
+  auto static cache = std::vector<int>{};
+  if (cache.size() == 0) {
+    auto device_count = int{};
+    raft_proto::cuda_check(cudaGetDeviceCount(&device_count));
+    cache.resize(device_count);
+    for (auto dev=0; dev < device_count; ++dev) {
+      raft_proto::cuda_check(
+        cudaDeviceGetAttribute(
+          &(cache[dev]),
+          cudaDevAttrMaxSharedMemoryPerMultiprocessor,
+          dev
+        )
+      );
+    }
+  }
+  return index_type(cache.at(device_id.value()));
+}
+
+inline auto get_mem_clock_rate(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
```

Comment:
What's the rationale behind caching some device attributes but not others?

Reply:
Profiling. Some of these calls are just barely expensive enough to justify caching, and some are faster than the necessary checking to retrieve from cache.